📖 Proposal for ClusterClass & managed topologies (#4678)
kubernetes-sigs/cluster-api: docs/proposals/202105256-cluster-class-and-managed-topologies.md

                kind: KubeadmConfigTemplate
                name: existing-boot-ref
            infrastructure:
              ref:
                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
                kind: VSphereMachineTemplate
                name: linux-vsphere-template
        - class: windows-worker
          template:
            bootstrap:
              ref:
                apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
                kind: KubeadmConfigTemplate
                name: existing-boot-ref-windows
            infrastructure:
              ref:
                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
                kind: VSphereMachineTemplate
                name: windows-vsphere-template
      infrastructure:
        ref:
          apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
          kind: VSphereClusterTemplate
          name: vsphere-prod-cluster-template
   ```
2. User creates a cluster using the class name and defining the managed topology.
   ```yaml
    apiVersion: cluster.x-k8s.io/v1alpha4
    kind: Cluster
    metadata:
      name: foo
      namespace: bar
    spec:
      topology:
        class: mixed

>         version: v1.19.1

Reviewer: Should we consider as part of the upgrade strategy talks on the ClusterClass side to allow folks to specify a minimum version that would then be rolled out to all clusters?

Author: Is this necessary for adopting existing clusters (which do not use a CC), since the `version` field is mandatory for creating new clusters with CC.

  Spec ClusterClassSpec `json:"spec,omitempty"`
}

// ClusterClassSpec describes the desired state of the ClusterClass.
type ClusterClassSpec struct {
  // ControlPlane is a reference to a local struct that holds the details
  // for provisioning the Control Plane for the Cluster.
  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`

  // Worker describes the worker nodes for the cluster.
  // It is a collection of node types which can be used to create
  // the worker nodes of the cluster.
  // +optional
  Worker WorkerClass `json:"worker,omitempty"`

  // Infrastructure is a reference to a provider-specific template that holds
  // the details for provisioning infrastructure specific cluster
  // for the underlying provider.
  // The underlying provider is responsible for the implementation
  // of the template to an infrastructure cluster.
  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
}

// WorkerClass is a collection of deployment classes which pronounce
//
type WorkerClass struct {
  // Deployments is a list of machine deployment types that can be used to create
  // a set of worker plane nodes.

>   Deployments []DeploymentClass `json:"deployments,omitempty"`

Reviewer 2: I'd rather avoid using the term deployment/deploymentClass since there's already a kubernetes resource with that same name.

Reviewer 3: MachineDeploymentClass??
But we can eventually reiterate during implementation and update the proposal accordingly

  // Class denotes a type of worker node present in the cluster
  Class string `json:"class,omitempty"`

  // Template is a local struct containing a collection of templates for creation of
  // MachineDeployment objects representing a set of worker nodes.
  Template DeploymentTemplateClass `json:"template"`
}

type DeploymentTemplateClass struct {
  Metadata ObjectMeta `json:"metadata,omitempty"`

  // Bootstrap contains the bootstrap template references to be used
  // for the creation of worker Machines.
  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`

  // Infrastructure contains the infrastructure template references to be used
  // for the creation of worker Machines.
  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
}

type LocalObjectTemplate struct {
  // Ref is a required reference to a custom resource
  // offered by a bootstrap provider.
  Ref corev1.ObjectReference `json:"ref"`
}
```

#### Modification to existing API Types
##### Cluster
1.  Add `Cluster.Spec.Topology` defined as
    ```golang
    // This encapsulates the topology for the cluster.
    // +optional
    Topology *Topology `json:"topology,omitempty"`
    ```

> 1.  The `ManagedTopology` object has the following definition:

Reviewer 4: ```suggestion
1.  The `Topology` object has the following definition:
```

    1. Adding/Removing MachineDeployment if necessary.
    1. Comparing and updating the number of replicas, if necessary.
    1. Comparing and updating the k8s version for the MD, if necessary.
    1. Updating the Machine Deployment object in the API server.

![Update cluster with ClusterClass][./images/cluster-class/update.png]

#### Provider implementation
##### For infrastructure providers
With the introduction of ClusterClass, we are extending the responsibility of the Cluster controller to create/update the actual objects from the collection of templates. Following this pattern, a template is required to create the actual infrastructure cluster object as part of the topology. For starter, the template should consist the InfraClusterSpec object. It can be enhanced to include other fields such as labels/annotations in the future.

The proposal calls for the implementation of an infrastructure cluster template to be referenced by the `ClusterClass` object. This is a required field for the `ClusterClass` object. This template would be used to create an infrastructure cluster object. Currently, this responsibility is assumed by the cluster operator/end user.

> **Note:** As per this proposal, the definition of ClusterClass is immutable. The CC definition consists of infrastructure object references, say AWSMachineTemplate, which could be immutable. For such immutable infrastructure objects, hard-coding the image identifiers leads to those templates being tied to a particular kubernetes version, thus making kubernetes version upgrades impossible. Hence, when using CC, infrastructure objects MUST NOT hard-code image identifiers.

Reviewer 5: ```suggestion
**Note:** As per this proposal, the definition of ClusterClass is immutable. The CC definition consists of infrastructure object references, say AWSMachineTemplate, which could be immutable. For such immutable infrastructure objects, hard-coding the image identifiers leads to those templates being tied to a particular kubernetes version, thus making kubernetes version upgrades impossible. Hence, when using CC, infrastructure objects MUST NOT have fields whose values prohibit version upgrades.
```

This method of provisioning the cluster would act as a single control point for the entire cluster. Scaling the nodes, adding/removing node pools and upgrading cluster kubernetes versions would be achievable by editing the managed topology. This would facilitate the maintenance of existing clusters as well as ease the creation of newer clusters.

### Goals

- Create the new ClusterClass CRD which can serve as a collection of templates to create clusters.
- Extend the Cluster object to use ClusterClass for creating managed topologies.
- Enhance the Cluster object to act as a single point of control for the managed topology.
- Extend the Cluster controller to create/update/delete managed topologies.

### Prospective future Work

⚠️ The following points are mostly ideas and can change at any given time  ⚠️

We are fully aware that in order to exploit the potential of ClusterClass and managed topologies, the following class of problems still needs to be addressed:

> - **Lifecycle of the ClusterClass**: Introduce mechanisms for allowing mutation of a ClusterClass, and the continuous reconciliation of the Cluster managed resources.

Reviewer 6: There might be a better place to put this, but I think it's worth having somewhere in the proposal.

```suggestion
- **Lifecycle of the ClusterClass**: Introduce mechanisms for allowing mutation of a ClusterClass, and the continuous reconciliation of the Cluster managed resources.
  - This effectively means that the only way to trigger an update of the ControlPlane and MachineDeployments is to change the Kubernetes version in the ClusterClass.
  - The ClusterClass defines Infrastructure References for the ControlPlane and MachineDeployments. These references are immutable, and the objects to which they refer (e.g. `AWSMachineTemplate` with CAPA) are also immutable. These objects must not hard-code machine image identifiers, since the identifier cannot be updated, and the machine image is bound to one Kubernetes patch version.
```

Author: > This effectively means that the only way to trigger an update of the ControlPlane and MachineDeployments is to change the Kubernetes version in the ClusterClass.

Updating the number of replicas would also result in updating the ControlPlane and MachineDeployments. Similarly, changing the labels/annotations would also result in an update. I guess you were specifically referring to updating the version in this case.

Reviewer 6: > > This effectively means that the only way to trigger an update of the ControlPlane and MachineDeployments is to change the Kubernetes version in the ClusterClass.
> 
> Updating the number of replicas would also result in updating the ControlPlane and MachineDeployments. Similarly, changing the labels/annotations would also result in an update. I guess you were specifically referring to updating the version in this case.

Good point. Reading it again, I think the above bullet is superfluous. I'll remove it. And I'll reword the second bullet a bit.

Author: @dlipovetsky The second bullet seems more of an implementation detail, I am not sure if including this in the proposal itself makes sense. The best place for this to reside is the v1alpha4 changes pages once the implementation is completed. Since this is something that the providers need to be aware of before they can start consuming the CC itself.

  Template TemplateClass `json:"template"`
}

type TemplateClass struct {
  Metadata ObjectMeta `json:"metadata,omitempty"`

  // Bootstrap contains the bootstrap template references to be used
  // for the creation of worker Machines.
  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`

  // Infrastructure contains the infrastructure template references to be used
  // for the creation of worker Machines.
  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
}

type LocalObjectTemplate struct {
  // Ref is a required reference to a custom resource
  // offered by a bootstrap provider.
  Ref corev1.ObjectReference `json:"ref"`
}
```

#### Modification to existing API Types
##### Cluster
1.  Add `Cluster.Spec.Topology` defined as
    ```golang
    // This encapsulates the managed topology for the cluster.
    // +optional
    Topology *ManagedTopology `json:"topology,omitempty"`
    ```
1.  The `ManagedTopology` object has the following definition:
    ```golang
    // ManagedTopology encapsulates the information of the managed resources.
    type ManagedTopology struct {
      // The name of the ClusterClass object to create the managed topology.

>       Class string `json:"class,omitempty"`

Reviewer 2: Does a managed topology have a status?
If not how do we know e.g in which version the cluster is?
Did it reconcile successfully last time spec.Version was changed?
What is successfully, is it the controlPlane upgraded or does it include all workers?

If we don't want to answer all this questions about upgrades we can may be drop the version field for now and add in a follow up with fleshed details or if we want to keep it we should add an upgrade caveat section or so capturing all these unknowns.

Reviewer: These all are valid concerns, longer term we can probably mirror some of these fields as a status field, although given that these APIs are not set in stone yet we can add conditions over time to signal.

A lot of these questions can be answered once the technical implementation is completed, given that we need to understand how the system works to properly signal successful version upgrades and possible caveats. We can amend the proposal along the way, although I wouldn't necessarily want to remove the version field now, but rather add more informational bits once we're ready to commit to some specific behaviors within the managed topology functionality.

Author: @enxebre Is this an acceptable path to move forward and retrofit the proposal with more details around this part?

      // The kubernetes version of the cluster.
      Version string `json:"version"`

      // RolloutAfter performs a rollout of the entire cluster one component at a time [...]
      // +optional
      RolloutAfter *metav1.Time `json:"rolloutAfter,omitempty"`

      // The information for the Control plane of the cluster.
      ControlPlane ControlPlane `json:"controlPlane"`

      // Worker encapsulates the different constructs that form the worker nodes
      // for the cluster.
      // +optional
      Worker *Worker `json:"worker,omitempty"`
    }
    ```
1.  The `ControlPlane` object contains the parameters for the control plane nodes of the managed topology.
    ```golang
    // ControlPlane specifies the parameters for the control plane nodes in the cluster.
    type ControlPlane struct {

>      Metadata ObjectMeta `json:"metadata,omitempty"`

Reviewer 7: Does a full ObjectMeta makes sense here or only a subset? (similar in other cases like Worker)

Author: This `ObjectMeta` struct is the CAPI struct defined in `v1a4/common_types.go`. It just contains the `Labels` and `Annotations` both of which are being used. I think you have it confused with the `ObjectMeta` from api-machinery. Hopefully, that resolves it.

       ```
      Note: The managed topology label needs to be set on the individual Machine objects as well.
    1. Creates the Machine Deployment object in the API server.

![Creation of cluster with ClusterClass][./images/cluster-class/create.png]

##### Update an existing Cluster using ClusterClass
This section talks about updating a cluster which was created using a `ClusterClass` object.
1. User updates the `cluster.spec.topology` field adhering to the update validation [criteria](#clusterclass-2).
2. For the ControlPlane object in `spec.topology.controlPlane`, the cluster controller checks for the presence of the control plane object using the name `<cluster-name>`. If found,
    1. Compares and updates the number of replicas, if necessary.
    1. Compares and updates the k8s version, if necessary.
    1. Updates the KCP object in the API server.
3. The cluster controller reconciles the list of required machine deployments with the current list of managed machine deployments by:
    1. Adding/Removing MachineDeployment if necessary.
    1. Comparing and updating the number of replicas, if necessary.
    1. Comparing and updating the k8s version for the MD, if necessary.

>     1. Updating the Machine Deployment object in the API server.

Reviewer 6: Let's say that the cluster has a MachineDeployment that uses a specific machine image. Let's say that I want to update the MachineDeployment to use a different machine image, e.g., because I need to use a more recent OS version, where a CVE has been patched. However, I cannot upgrade the cluster's Kubernetes version, because I am on the latest available version.

How do I update the MachineDeployment's machine image?

If I do not use ClusterClass, then I create a new Infrastructure Template (e.g. AWSMachineTemplate if I'm using CAPA), and update the MachineDeployment's Infrastructure Reference. 

Assuming I understand the proposal correctly, if use ClusterClass, I cannot update the MachineDeployment's Infrastructure Reference, because ClusterClass defines the Reference, and the definition is immutable.

Author: Yes, that assumption is correct. This version of ClusterClass does not allow changes to the ClusterClass apart from adding new worker sets. As part of the future work, we have added a goal to allow cluster-specific transformations to change the values present in the ClusterClass. Once the baseline CC proposal is implemented, the plan is to move towards introducing these mutability options while defining the managed topology in the cluster.

            - [ClusterClass](#clusterclass-2)
            - [Cluster](#cluster-1)
        - [Behaviors](#behaviors)
            - [Create a new Cluster using ClusterClass object](#create-a-new-cluster-using-clusterclass-object)
            - [Update an existing Cluster using ClusterClass](#update-an-existing-cluster-using-clusterclass)
        - [Provider implementation](#provider-implementation)
            - [For infrastructure providers](#for-infrastructure-providers)
            - [For Control plane providers](#for-control-plane-providers)
      - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
      - [Test Plan [optional]](#test-plan-optional)
      - [Graduation Criteria [optional]](#graduation-criteria-optional)
      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
  - [Implementation History](#implementation-history)

## Glossary

### ClusterClass
A collection of templates that define a topology (control plane and machine deployments) to be used in one or more Managed Clusters.

### Managed Cluster topology

> A Managed Cluster is a Cluster that provides a single control point to manage it's own topology; the managed cluster topology is defined by a ClusterClass.

Reviewer 4: ```suggestion
A managed cluster topology refers to a Cluster that provides a single control point to manage its own topology; the managed cluster topology is defined by a ClusterClass.
```

> 

Reviewer 2: Let's be consistent with above -> ClusterClass and WorkerClass

> ### Worker class

Reviewer 2: So is a `clusterClass` composed of 0..N `workerClass`? if so probably worth call that out in this glossary.

Author: It is not just a collection of worker classes but also defines the control plane template. I can add something similar to it though.

A worker class is a collection of templates that define a set of worker nodes in the cluster.

## Summary

This proposal introduces a new ClusterClass object which will be used to provide easy stamping of clusters of similar shapes. It serves as a collection of template resources which are used to generate one or more clusters of the same flavor.

We're enhancing the Cluster CRD and controller to use a ClusterClass resource to provision the underlying objects that compose a cluster. Additionally, the Cluster provides a single control point to manage the Kubernetes version, worker pools, labels, replicas, and so on.

## Motivation

Currently, Cluster API does not expose a native way to provision multiple clusters of the same configuration. The ClusterClass object is supposed to act as a collection of template references which can be used to create managed topologies.

> Today, the Cluster object is a logical grouping of components which describe an underlying cluster. The user experience to create a cluster requires the user to create a bunch of underlying resources such as KCP (control plane provider), MachineDeployments, and infrastructure or bootstrap templates for those resources which logically end up representing the cluster. Since the cluster configuration is spread around multiple components, upgrading the cluster version is hard as it requires changes to different fields in different resources to perform an upgrade.

Reviewer 2: Can we elaborate on why is a clusterClass necessary or the solution choosen to have a single point to manage version?

This method of provisioning the cluster would act as a single control point for the entire cluster. Scaling the nodes, adding/removing node pools and upgrading cluster kubernetes versions would be achievable by editing the managed topology. This would facilitate the maintenance of existing clusters as well as ease the creation of newer clusters.

### Goals

- Create the new ClusterClass CRD which can serve as a collection of templates to create clusters.
- Extend the Cluster object to use ClusterClass for creating  managed cluster topologies.
- Enhance the Cluster object to act as a single point of control for the managed topology.

> - Extend the Cluster controller to create/update/delete managed topologies.

Reviewer 2: This sounds the same than the future work below:
`Introduce mechanisms for allowing mutation of a ClusterClass, and the continuous reconciliation of the Cluster managed resources.`

So may be update is not part of the goal?

Reviewer 8: I think what's meant here is that the `ClusterSpec` will be extended by a `managed` field, which contains the reference to a ClusterClass and config like the amount of nodes.

In order for ClusterClass to  be usable, the Cluster controller needs to reconcile that new field and take care of creating the resources as specified in the ClusterClass.

Author: Like @schrej mentioned above, the update is to handle the config updates to `spec.Managed` field.

place your `.plantuml` files under `images/` and run `make diagrams` from the docs folder.

### User Stories

#### Story 1 - Use ClusterClass to easily stamp clusters
As an end user, I want to use one `ClusterClass` to create multiple managed cluster topologies of similar flavor.
- Rather than recreating the KCP and MD objects for every cluster that needs to be provisioned, the end user can create a template once and reuse it to create multiple clusters with similar configurations.

#### Story 2 - Easier UX for kubernetes version upgrades
For an end user, the UX to update the kubernetes version of the control plane and worker nodes in the cluster should be easy.
- Instead of individually modifying the KCP and each MachineDeployment, updating a single option should result in k8s version updates for all the CP and worker nodes.

**Note**: In order to complete the user story for all the providers, some of the advanced features (such as Extensibility/Transformation) are required. However, getting this in place even only for a subset of providers allows us to build and test a big chunk of the entire machinery.

#### Story 3 - Easier UX for scaling workers nodes

> As an end user, I want to be able to easily scale up/down the number of replicas for each set of worker nodes in the cluster.

Reviewer 2: Can we elaborate on the real use case for this UX? e.g is this thinking on having a single point fo scaling machines evenly across AZs? Is it something else?

Reviewer 8: Instead of having to modify one or more MachineDeployments, you can simply change the desired worker count of the Cluster resource. If they're managed that's also the only way, as you probably shouldn't touch the resources created by the controller.

Reviewer 2: >Instead of having to modify one or more MachineDeployments, you can simply change the desired worker count of the Cluster resource. If they're managed that's also the only way, as you probably shouldn't touch the resources created by the controller.

Make sense so It'd be nice to illustrate a bit more here user story/scenarios/topologies where this UX and functionality (ability to scale multiple scalable resources in a single CR) will come in handy rather than describing the functionality itself. E.g for the opposite use case: As a cluster operator I want to provide certain users (e.g Autoscaler, Tester...) with granular perms to scale machines but prevent them from modifying cluster properties so I want this ability to be scoped in different CRs.

>As I'm not familiar how CAPI handles AZs this is just a guess, but since the managed workers are also created using a MachineDeployment, which is created from some form of template, I guess it will continue to depend on the provider in use how distributing nodes across AZs has to be configured.

Nothing does that for you today, each pool of nodes for a machineDeployment belong to a single AZ unless you are using MachinePools. The functionality described here might ease that manual task for an operator, E.g As a cluster operator I want to have a single point to scale my MachineDeployments so I can easily ensure they result in an even number of workers across AZs to be more tolerant to failures.

If instead you are eager to see an example of ClusterClass and how the Cluster object will look like, you can jump to the Behaviour paragraph.

#### New API types
##### ClusterClass
This CRD is a collection of templates that describe the topology of one or more managed clusters.
```golang
// +kubebuilder:object:root=true
// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
// +kubebuilder:storageversion

// ClusterClass is a template which can be used to create managed cluster topologies.
type ClusterClass struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`

  Spec ClusterClassSpec `json:"spec,omitempty"`
}

// ClusterClassSpec describes the desired state of the ClusterClass.
type ClusterClassSpec struct {
  // ControlPlane is a reference to a local struct that holds the details
  // for provisioning the Control Plane for the Cluster.
  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`

  // Worker describes the worker nodes for the cluster.
  // It is a collection of node types which can be used to create
  // the worker nodes of the cluster.
  // +optional

>   Worker WorkerClass `json:"worker,omitempty"`

Reviewer 2: why not name this workerClass?

Reviewer 8: I assume the struct was called WorkerClass here since it's part of ClusterClass, and Worker is used in the extension to the Cluster resource.

Author: The idea was to clearly demarcate the ClusterClassSpec into:
- Control Plane node definition
- Worker node definition

Adding the `Class` prefix would make it appear as `workerClass.Deployments[0].Class` which makes the naming scheme slightly confusing.

Author: Copying over a discussion from the Google doc which is about WorkerClass struct too

> Puja Abbassi
> 4:47 AM Jun 1
> Is this an "3 similar node pool each in a different AZ" use case only? Otherwise, I'm not sure I've seen end users scaling their node pools within a single cluster in the same way, especially when the node pools have different instance/machine sizes. 
> Maybe for multi-tenant clusters that have a node pool per team that could be a use case, but tbh I've not seen this use case a lot.
> I can however think of use cases where the same node pool in different clusters should be scaled similarly.
> Most probably as long as the WorkerClass can be used within and across clusters we can serve both types of use cases.
> 
> Sagar Muchhal
> 10:45 AM Jun 1
> The intent of the example was to showcase that currently to scale different sets of worker nodes, we need to update the MD objects representing those sets of worker nodes.
> Instead of going through multiple object updates, we can individually update the replica number for each of these worker sets under the `spec.Managed` field and thereby just having to update a single object once (with diff number of replicas for each set of worker nodes).
> Hopefully this resolves any confusion that the current wording may have caused.
> 
> Puja Abbassi
> 11:26 AM Jun 1
> I was just a bit confused, cause above there's a mention of WorkerClass and I thought that would be a separate object, but below I do not see it anymore, or rather workers are defined within a CC. The example you describe makes sense in the context of managing sets of workers within a single cluster.I was just thinking that there are similarly use cases for managing sets of workers across different clusters that have similar needs but might have differences on the configuration of the CC. That made me wonder if having worker class and cluster class separated could be valuable to create a more modular approach to managing clusters. Is that out of scope here?
> 
> Sagar Muchhal
> 11:57 AM Jun 1
> That is an interesting thought, of having separate class types to individually control the CP and worker nodes, and making the CC a collection of these classes. I would like to hear more from the community about what they think about this usecase.
> The naming `WorkerClass` was just to keep the objects aligned with the concept that they belong to. While outlining the proposal, this wasn't a usecase we designed for. Plus inlining the definition of the WorkerClass object makes it easier to define a CC.

// provisioned using the `ClusterClass`.
type DeploymentClass struct {
  // Class denotes a type of worker node present in the cluster
  Class string `json:"class,omitempty"`

  // Template is a local struct containing a collection of templates for creation of
  // MachineDeployment objects representing a set of worker nodes.
  Template TemplateClass `json:"template"`
}

type TemplateClass struct {
  Metadata ObjectMeta `json:"metadata,omitempty"`

  // Bootstrap contains the bootstrap template references to be used
  // for the creation of worker Machines.
  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`

  // Infrastructure contains the infrastructure template references to be used
  // for the creation of worker Machines.
  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
}

type LocalObjectTemplate struct {
  // Ref is a required reference to a custom resource
  // offered by a bootstrap provider.
  Ref corev1.ObjectReference `json:"ref"`
}
```

#### Modification to existing API Types
##### Cluster
1.  Add `Cluster.Spec.Managed` defined as
    ```golang
    // The information of the managed resources for the cluster.
    // +optional

>     Managed *ManagedCluster `json:"managed,omitempty"`

Reviewer 4: ```suggestion
    Topology * ManagedTopology `json:"topology,omitempty"`
```

Author: I think the name of the field should still be `Managed` even if we update the name of the struct to `ManagedTopology`. I agree that the name **Managed Cluster** tends to be misleading, and **Managed Topology** seems to be an appropriate replacement. 

But the field named `Topology` does not convey that it is indeed managed by the Cluster object. Keeping the field name `Managed` does emphasize the managed aspect of the topology. wdyt?

Reviewer 9: I missed this comment as it was hidden by an outdated / resolved block in the UI. I've added another comment in a similar theme: https://github.com/kubernetes-sigs/cluster-api/pull/4678#pullrequestreview-677406313.

If there are not other types of topologies, then it seems odd to draw this distinction.

    ```
1.  The `ManagedCluster` object has the following definition:
    ```golang
    // ManagedCluster encapsulates the information of the managed resources.

>     type ManagedCluster struct {

Reviewer 4: ```suggestion
    type ManagedClusterTopology struct {
```

Reviewer 4: or 

```suggestion
    type ManagedTopology struct {
```

preference for the latter

      // The information for the Control plane of the cluster.
      ControlPlane ControlPlane `json:"controlPlane"`

      // Worker encapsulates the different constructs that form the worker nodes
      // for the cluster.
      // +optional
      Worker *Worker `json:"worker,omitempty"`
    }
    ```
1.  The `ControlPlane` object contains the parameters for the control plane nodes of the managed cluster.
    ```golang
    // ControlPlane specifies the parameters for the control plane nodes in the cluster.
    type ControlPlane struct {
     Metadata ObjectMeta `json:"metadata,omitempty"`

     // The number of control plane nodes.
     Replicas int `json:"replicas"`
    }
    ```
1.  The `Worker` object represents the sets of worker nodes of the managed cluster.
    ```golang
    // Worker represents the different sets of worker nodes in the cluster.
    type Worker struct {
      Deployments []Deployment `json:”deployments,omitempty”`
    }
    ```
1.  The `Deployment` object represents a single set of worker nodes of the managed cluster.
    ```golang
    // Deployment specifies the different parameters for a set of worker nodes in the cluster.

>     type Deployment struct {

Reviewer 2: We should probably avoid the term deployment to avoid confusion with appsv1.Deployment

Reviewer 8: yes

Author: The idea behind naming it as `Deployment` was to map it to `MachineDeployment`. And in the future, introduce `Pool` which would correspond to `MachinePool`.

      Metadata ObjectMeta `json:"metadata,omitempty"`

      // Class is the name of the DeploymentClass used to create the set of worker nodes.
      // This should match one of the deployment classes defined in the ClusterClass object
      // mentioned in the `Cluster.Spec.Class` field.
      Class string `json:"class"`

      // Name refers to the name of the set of worker nodes.
      Name string `json:"name"`

      // The number of worker nodes belonging to this set.
      // If not specified, the cluster autoscaler for CAPI would be responsible
      // for maintaining a min/max number of replicas.
      // +optional
      Replicas *int `json:"replicas,omitempty"`

>     }

Reviewer 2: Where in this picture is the workerClass used?

Reviewer 2: Got it, so https://github.com/kubernetes-sigs/cluster-api/pull/4678/files#r642478474

Reviewer 8: WorkerClass is not really a "class", it's just there to make ClusterClass easier to extend in the future, if that should be necessary

                kind: KubeadmConfigTemplate
                name: existing-boot-ref
            infrastructure:
              ref:
                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
                kind: VSphereMachineTemplate
                name: linux-vsphere-template
        - class: windows-worker
          template:
            bootstrap:
              ref:
                apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
                kind: KubeadmConfigTemplate
                name: existing-boot-ref-windows
            infrastructure:
              ref:
                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
                kind: VSphereMachineTemplate
                name: windows-vsphere-template
      infrastructure:
        ref:
          apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
          kind: VSphereClusterTemplate
          name: vsphere-prod-cluster-template
   ```
2. User creates a cluster using the class name and defining the managed topology.
   ```yaml
    apiVersion: cluster.x-k8s.io/v1alpha4
    kind: Cluster
    metadata:
      name: foo
      namespace: bar
    spec:

>       managed:

Reviewer 4: What about just calling it `topology:`? The topology itself is managed, not the cluster. Calling a cluster "managed" just implies other things and I really think we should stay away from that. `topology:` would imply, "this is what my cluster's topology should look like". 

Anyone else has suggestions?

cc @puja108

>       managed:

Reviewer 3: Naming is hard

What I like of the current name is that it reflects the fact that everything is under Cluster.Managed is actually controlled (aka managed) by the Cluster CR/Cluster controller in Cluster API.

Instead, considering that user is still free to add more MachineDeployments/Pools manually and self-manage them as of today, I find the use of the term topology a little bit trickier, because in my mind such mixed scenario requires some more nuances like: 
- The "cluster topology" including both MachineDeployments managed at Cluster level and the MachineDeployments/Pools manually added/self managed.
- The "managed topology" of a cluster, including only MachineDeployments managed at Cluster level
- The "self-managed topology" of a cluster, including only MachineDeployments/Pools manually added

However I agree, that the downside of "managed" is the potential confusion with AKS, EKS, GKE, which are Kubernetes offering managed by something/someone outside of Cluster API.

>       managed:

Reviewer 10: HUGE +1 from me as well. We really can't start referring to clusters built/defined with this new `ClusterClass` configuration by a different name (e.g., "managed cluster" vs "cluster"). In this particular case, the term "managed cluster" is too general, and in the cluster-api ecosystem it is very sensible right now to refer to *any* cluster-api workload cluster as a "managed cluster" — these cluster *are managed* by the cluster-api _management cluster_.

And, more broadly, this seems to be a natural evolution of the cluster-api abstraction. We should expect there to be many more in the future. We don't want to back ourselves into a corner by having to create a distinct terminology for every evolving "type" of cluster-api cluster configuration.

My 2¢ is that we should try to always refer to this in language as "a cluster that uses ClusterClass" to reflect that this is just one possible configuration paradigm for a cluster-api cluster, one that reflects a certain vintage ("I built my cluster from cluster-api version XYZ and am taking advantage of the latest functionality!") and operational preference. Rather than "a discrete _type_ of cluster-api cluster".

>       managed:

Author: I agree with the concern that the term **Managed Cluster** is already in use outside of the context of CAPI, and overloading the use of the term will be misleading. 
But does naming the field `managed` imply that the cluster is managed? IMO it just states that the spec described within this field is managed and not necessarily that the cluster is managed. The confusion could be arising since the golang struct is named as `ManagedCluster` and I vote for updating that name to `ManagedTopology` to alleviate the confusion and keeping the name of the field as `managed`.

        class: mixed
        version: v1.19.1
        controlPlane:
          replicas: 3
          labels: {}
          annotations: {}
        worker:
          deployments:
          - class: linux-worker
            name: big-pool-of-machines-1
            replicas: 5
            labels:
              # This label is additive to the class' labels,
              # or if the same label exists, it overwrites it.
              custom-label: "production"
          - class: linux-worker
            name: small-pool-of-machines-1
            replicas: 1
          - class: windows-worker
            name: microsoft-1
            replicas: 3
   ```
3. The Cluster controller checks for the presence of the `spec.managed.class` field. If the field is missing, the Cluster controller behaves in the existing way. Otherwise the controller starts the creation process of the managed topology defined in the steps below.
4. Cluster and Control plane object creation
    1. Creates the infrastructure specific/provider cluster using the cluster template referenced in the `ClusterClass.spec.infrastructure.ref` field.
    1. Add the managed topology label to the provider cluster object:
       ```yaml

>         cluster.x-k8s.io/managed:<cluster-name>

Author: I would say let's update the label for all objects under the managed topology for the sake of consistency.

>         cluster.x-k8s.io/managed:<cluster-name>

Author: Updated all the instances of this label.

       ```
    1. For the ControlPlane object in `cluster.spec.managed.controlPlane`
    1. Initializes a control plane object using the control plane template defined in the `ClusterClass.spec.controlPlane.ref field`. Use the name `<cluster-name>`.
    1. Sets the number of replicas on the control plane object from `spec.managed.controlPlane.replicas`.
    1. Sets the k8s version on the control plane object from the `spec.managed.version`.
    1. Add the following labels to the control plane object:
       ```yaml
        cluster.x-k8s.io/managed: <cluster-name>
       ```
    1. Creates the control plane object.
    1. Sets the `spec.controlPlaneRef` field for the Cluster object.
    1. Saves the cluster object in the API server.
5. Machine deployment object creation
    1. For each `spec.managed.worker.deployments` item in the list
    1. Create a name `<cluster-name>-<worker-set-name>` (if too long, hash it)
    1. Initializes a new MachineDeployment object.
    1. Sets the `clusterName` field on the MD object
    1. Sets the `replicas` field on the MD object using `replicas` field for the set of worker nodes.

>     1. Sets the `version` field on the MD object from the `spec.managed.version`.

Reviewer 8: We're using different machine templates depending on the k8s version, something that couldn't be done with what's described in this proposal. We can solve it with templating, which is planned for later, so it might not make sense to discuss it here, but I wanted to mention it anyway.

Maybe it would be an option to provide multiple DeploymentClasses and have an (optional) Version field in them and then select one based on version automatically?

        cluster.x-k8s.io/managed: <cluster-name>-<worker-set-name>
       ```
      Note: The managed topology label needs to be set on the individual Machine objects as well.
    1. Creates the Machine Deployment object in the API server.

![Creation of cluster with ClusterClass][./images/cluster-class/create.png]

##### Update an existing Cluster using ClusterClass
This section talks about updating a cluster which was created using a `ClusterClass` object.
1. User updates the `cluster.spec.managed` field adhering to the update validation [criteria](#clusterclass-2).
2. For the ControlPlane object in `spec.managed.controlPlane`, the cluster controller checks for the presence of the control plane object using the name `<cluster-name>`. If found,
    1. Compares and updates the number of replicas, if necessary.
    1. Compares and updates the k8s version, if necessary.
    1. Updates the KCP object in the API server.
3. The cluster controller reconciles the list of required machine deployments with the current list of managed machine deployments by:
    1. Adding/Removing MachineDeployment if necessary.
    1. Comparing and updating the number of replicas, if necessary.

>     1. Comparing and updating the k8s version for the MD, if necessary.

Reviewer 2: This would result in all machineDeployments being upgraded at the same time. Are we considering any kind of policy/strategy e.g upgrading only one at a time and only move to next one when the previous one has completed, rollbacks... I think the upgrading workflow probably deserves its own section.

Author: I like the suggestion of having a control strategy around handling MD upgrades if a config (say k8s version) is changed. In the short term, we can use sensible defaults on the MachineDeploymentStrategy to ensure that not all the worker nodes get updated at the same. Currently, the default strategy type is `RollingUpdate` and `MaxSurge` is 1. This will ensure that even though all the MDs are getting updated at the same time, the overall availability of the worker nodes is not hampered.

Providing the observability metrics of the cluster status is outlined as a future goal and I suppose we can tackle the state of MD upgrades (yet to be started/in progress/completed) as a part of that enhancement.

Author: Another idea to control the update/rollback strategy could be to expose the `MachineDeploymentStrategy` field of the MD object under the `Managed` section of the Cluster object thereby not introducing a new level of complexity to these updates.

Reviewer: We definitely need a rollout strategy for the class itself, although that can be added later

Author: Added a **Future goal** to implement an upgrade/rollback strategy.

    1. Updating the Machine Deployment object in the API server.

![Update cluster with ClusterClass][./images/cluster-class/update.png]

#### Provider implementation
##### For infrastructure providers

> The proposal also calls for the implementation of an infrastructure cluster template to be referenced by the `ClusterClass` object to create an infrastructure cluster object for a CAPI cluster. This template is a required field to create a `ClusterClass`.

Author: Added some background information and expectation of this new Template.

##### For Control plane providers
A control plane provider should create a template for use of creation for the control plane objects. For instance, the kubeadm control plane provider should introduce the notion of a KubeadmControlPlaneTemplate for use by the cluster controller to create a managed Kubeadm Control plane object. The minimum set of parameters supported by the control plane template are:

>   - infrastructure template reference

Author: Updated the section.

### Risks and Mitigations

This proposal tries to model the API design for ClusterClass with a narrow set of usecases. This initial implementation provides a baseline on which inceremental changes can be introduced in the future. Instead of encompassing of all usecases under a single proposal, this proposal mitigates the risk of waiting too long to consider all required usecases under this topic.

> ## Alternatives

Reviewer 11: I guess use different interfaces/resources for different uses cases. For example, I see a clear distinction between `ClusterTemplate` which will hold the configuration aspect of a cluster to let users create clusters with the same values, and for other part something like `ClusterSet` which will allow you to coordinate/manage actions through a set of clusters like upgrades, scaling,...

> ## Alternatives

Author: Removed the section, I misunderstood the intent behind it.

        - [Validations](#validations)
            - [ClusterClass](#clusterclass-2)
            - [Cluster](#cluster-1)
        - [Behaviors](#behaviors)
            - [Create a new Cluster using ClusterClass object](#create-a-new-cluster-using-clusterclass-object)
            - [Update an existing Cluster using ClusterClass](#update-an-existing-cluster-using-clusterclass)
        - [Provider implementation](#provider-implementation)
            - [For infrastructure providers](#for-infrastructure-providers)
            - [For Control plane providers](#for-control-plane-providers)
      - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
      - [Test Plan [optional]](#test-plan-optional)
      - [Graduation Criteria [optional]](#graduation-criteria-optional)
      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
  - [Implementation History](#implementation-history)

## Glossary

### ClusterClass
A collection of templates that define a topology (control plane and machine deployments) to be used in one or more clusters.

> ### Managed topology

Reviewer 9: Found a similar themed comment: https://github.com/kubernetes-sigs/cluster-api/pull/4678#discussion_r644375289

> ### Managed topology

Author: The term `Managed` explicitly emphasizes that the topology bits are managed by the Controller. On the other hand, clusters without CC have their topology bits managed by the operators. The term topology could be used to describe the cluster structure but the prefix `Managed` just makes the intent explicit.

The only other modifier that I can think of is **Supervised** but I still feel it does not correctly describe the intent.

> ### Managed topology

Reviewer 9: To follow the same logic, do we call a `Machine` a `ManagedMachine` because the bits are managed by the Controller?

> ### Managed topology

Author: This particular scenario is different wherein a different set of CRDs are managed by a controller which isn't primarily responsible for those ones. Hence the extra emphasis on the managed aspect.

> ### Managed topology

Reviewer 9: This may have some deeper meaning for someone really close to the project, but to an outsider, this will be a difference without a distinction. 

Again, will there be other types of Topologies? Will there be UnmanagedTopologies?

> ### Managed topology

Author: There is no concept of `UnmanagedTopology` in the works or planned in the future. I am ok to drop the **Managed** prefix in the name and add the managed aspect to the definition of the topology itself.

  // the worker nodes of the cluster.
  // +optional
  Worker WorkerClass `json:"worker,omitempty"`

  // Infrastructure is a reference to a provider-specific template that holds
  // the details for provisioning infrastructure specific cluster
  // for the underlying provider.
  // The underlying provider is responsible for the implementation
  // of the template to an infrastructure cluster.
  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
}

// WorkerClass is a collection of deployment classes which pronounce
//
type WorkerClass struct {
  // Deployments is a list of machine deployment types that can be used to create
  // a set of worker plane nodes.
  Deployments []DeploymentClass `json:"deployments,omitempty"`
}

// DeploymentClass serves as a template to define a set of worker nodes of the cluster
// provisioned using the `ClusterClass`.
type DeploymentClass struct {
  // Class denotes a type of worker node present in the cluster
  Class string `json:"class,omitempty"`

  // Template is a local struct containing a collection of templates for creation of
  // MachineDeployment objects representing a set of worker nodes.

>   Template TemplateClass `json:"template"`

Reviewer 2: I'd expect TemplateClass to have a naming reference to what's templating, e.g `MachineTemplateClass`

Author: How about `DeploymentTemplateClass` instead, since it is a collection of templates for the MachineDeployment?

Reviewer 2: Make sense, though I'd be also in favour of not using the term Deployment here either since there's already a core API CR with that name, any reason why we are using deployment rather than machineDeployment? e.g
```
    type Worker struct {
      MachineDeployments []ManagedMachineDeployment `json:”machineDeployments,omitempty”`
    }
```

Or if we don't want to couple it to this particular scalable resource may be:

```
    type Worker struct {
      nodePool []ManagedNodePool `json:”machineDeployments,omitempty”`
    }
```
or something similar?

Author: Changed :+1:

      // This should match one of the deployment classes defined in the ClusterClass object
      // mentioned in the `Cluster.Spec.Class` field.
      Class string `json:"class"`

      // Name refers to the name of the set of worker nodes.
      Name string `json:"name"`

      // The number of worker nodes belonging to this set.
      // If not specified, the cluster autoscaler for CAPI would be responsible
      // for maintaining a min/max number of replicas.
      // +optional
      Replicas *int `json:"replicas,omitempty"`
    }
    ```

#### Validations
##### ClusterClass
- For object creation:
  - `spec.worker.deployments[i].class` field must be unique within a ClusterClass.
- For object updates:
  - `spec.worker.deployments` supports adding new deployment classes.

##### Cluster
- For object creation:
  - `spec.topology` and `spec.controlPlaneRef` cannot be simultaneously set.
  - If `spec.topology.class` is set, `spec.topology.version` cannot be empty and must be a valid semver.

- For object updates:
  - If spec.topology.class is in use, it cannot be unset or modified.
  - spec.topology.version cannot be unset and must be a valid semver, if being updated.

>   - `spec.topology.version` cannot be downgraded.

Reviewer 2: How do we support rollbacks for scenarios where machineDeployments get stuck during an upgrade?

Author: Added a **Future goal** to implement an upgrade/rollback strategy.

Author: Similar to the above comment around version upgrades.

CAPI still supports the default cluster creation functionality wherein the users can supply the individual object definitions (KCP, MD, etc.) which make up a cluster.

## Upgrade Strategy

Existing clusters created without ClusterClass cannot switch over to using ClusterClass for a managed topology.

## Additional Details

### Test Plan [optional]

**Note:** *Section not required until targeted at a release.*

Consider the following in developing a test plan for this enhancement:
- Will there be e2e and integration tests, in addition to unit tests?
- How will it be tested in isolation vs with other components?

No need to outline all of the test cases, just the general strategy.
Anything that would count as tricky in the implementation and anything particularly challenging to test should be called out.

All code is expected to have adequate tests (eventually with coverage expectations).
Please adhere to the [Kubernetes testing guidelines][testing-guidelines] when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md

> ### Graduation Criteria [optional]

Reviewer 2: Is this intending to go under a gate flag initially?

Author: The intention is to ship it under a feature gate initially. Will add that to the proposal.

- Enhance the Cluster object to act as a single point of control for the managed topology.
- Extend the Cluster controller to create/update/delete managed topologies.

### Prospective future Work

⚠️ The following points are mostly ideas and can change at any given time  ⚠️

We are fully aware that in order to exploit the potential of ClusterClass and managed topologies, the following class of problems still needs to be addressed:
- **Lifecycle of the ClusterClass**: Introduce mechanisms for allowing mutation of a ClusterClass, and the continuous reconciliation of the Cluster managed resources.
- **Extensibility/Transformation**: Introduce mechanism for allowing Cluster specific transformations of a ClusterClass (e.g. inject API Endpoint for CAPV, customize machine image by version etc.)
- **Adoption**: Providing a way to convert existing clusters into managed cluster topologies.
- **Observability**: Build an SDK and enhance the Cluster object status to surface a summary of the status of managed topology.
- **Healthchecks**: Augment ClusterClass to include MachineHealthCheck refs to manage the health of the managed topology.

> - **Lifecycle integrations**: Extend ClusterClass to include lifecycle management integrations such as MahcineHealthCheck and Cluster Autoscaler to manage the state and health of the managed clusters.

Reviewer 4: ```suggestion
- **Lifecycle integrations**: Extend ClusterClass to include lifecycle management integrations such as MachineHealthCheck and Cluster Autoscaler to manage the state and health of the managed clusters.
```
